File: cpp/use_vertical_formatting.cpp
// Use Vertical Formatting
// You should read your code from top to bottom.
// You should read your code without "jumping" over functions.
// Similar and dependent functions should be close vertically.

#include <iostream>
#include <vector>

// Before
class CalculatorBefore {
public:
    int add(int a, int b) {
        return a + b;
    }

    int multiply(int a, int b) {
        return a * b;
    }

    std::vector<int> calculate(int a, int b) {
        int sumResult = add(a, b);
        int difference = subtract(a, b);
        int product = multiply(a, b);
        int quotient = divide(a, b);
        return {sumResult, difference, product, quotient};
    }

    int divide(int a, int b) {
        return a / b;
    }

    int subtract(int a, int b) {
        return a - b;
    }
};

// After
class Calculator {
public:
    std::vector<int> calculate(int a, int b) {
        int sumResult = add(a, b);
        int difference = subtract(a, b);
        int product = multiply(a, b);
        int quotient = divide(a, b);
        return {sumResult, difference, product, quotient};
    }

    int add(int a, int b) {
        return a + b;
    }

    int subtract(int a, int b) {
        return a - b;
    }

    int multiply(int a, int b) {
        return a * b;
    }

    int divide(int a, int b) {
        return a / b;
    }
};

void printVector(const std::vector<int>& vec) {
    std::cout << "[";
    for (size_t i = 0; i < vec.size(); ++i) {
        std::cout << vec[i];
        if (i < vec.size() - 1) std::cout << ", ";
    }
    std::cout << "]" << std::endl;
}

int main() {
    CalculatorBefore calculatorBefore;
    auto resultBefore = calculatorBefore.calculate(10, 5);
    printVector(resultBefore); // [15, 5, 50, 2]

    std::cout << std::endl;

    Calculator calculator;
    auto result = calculator.calculate(10, 5);
    printVector(result); // [15, 5, 50, 2]

    return 0;
}

File: rust/use_vertical_formatting.rs
// Use Vertical Formatting
// You should read your code from top to bottom.
// You should read your code without "jumping" over functions.
// Similar and dependent functions should be close vertically.

// Before
struct CalculatorBefore;

impl CalculatorBefore {
    fn add(&self, a: i32, b: i32) -> i32 {
        a + b
    }

    fn multiply(&self, a: i32, b: i32) -> i32 {
        a * b
    }

    fn calculate(&self, a: i32, b: i32) -> Vec<i32> {
        let sum_result = self.add(a, b);
        let difference = self.subtract(a, b);
        let product = self.multiply(a, b);
        let quotient = self.divide(a, b);
        vec![sum_result, difference, product, quotient]
    }

    fn divide(&self, a: i32, b: i32) -> i32 {
        a / b
    }

    fn subtract(&self, a: i32, b: i32) -> i32 {
        a - b
    }
}

// After
struct Calculator;

impl Calculator {
    fn calculate(&self, a: i32, b: i32) -> Vec<i32> {
        let sum_result = self.add(a, b);
        let difference = self.subtract(a, b);
        let product = self.multiply(a, b);
        let quotient = self.divide(a, b);
        vec![sum_result, difference, product, quotient]
    }

    fn add(&self, a: i32, b: i32) -> i32 {
        a + b
    }

    fn subtract(&self, a: i32, b: i32) -> i32 {
        a - b
    }

    fn multiply(&self, a: i32, b: i32) -> i32 {
        a * b
    }

    fn divide(&self, a: i32, b: i32) -> i32 {
        a / b
    }
}

fn main() {
    let calculator_before = CalculatorBefore;
    let result_before = calculator_before.calculate(10, 5);
    println!("{:?}", result_before); // [15, 5, 50, 2]

    println!();

    let calculator = Calculator;
    let result = calculator.calculate(10, 5);
    println!("{:?}", result); // [15, 5, 50, 2]
}
